<h1><center>Лабораторная работа №1 "Библиотека Pandas".</center></h1>
<h1><center>Вахлаева Марина Олеговна</center></h1>

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

<h3><center>1. Считать в pandas.DataFrame любой источник данных: CSV, JSON, Excel-файл, HTML-таблицу и т.п. Также можно сконвертировать в DataFrame любой из встроенных датасетов sklearn.</center></h3>

In [2]:
path = 'C:/Users/PC/Downloads/Python/TAI/datasets/'

df = pd.read_csv(Path(path, 'heart.csv'))

In [3]:
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

    ca  thal  target  
0  0.0   1.0       1  
1  0.0   2.0       1  
2  0.0   2.0       1  
3  0.0   2.0       1  
4  0.0   2.0       1

In [4]:
type(df)

pandas.core.frame.DataFrame

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        298 non-null    float64
 12  thal      301 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


<h3><center>2. Привести описание датасета.</center></h3>

Датасет содержит данные 303 пациентов, которые склонны к сердечно-сосудистым заболеваниям. Всего было взято 14 показателей: 
- возраст; 
- пол (1 - мужчина, 0 - женщина); 
- "cp" = тип боли в груди (0 - типичная стенокардия, 1 - атипичная стенокардия, 2 - неангинальная боль, 3 - бессимптомное); 
- "trestbps" = кровяное давление в состоянии покоя (в мм рт. ст. при поступлении в больницу);
- "chol" = холестерин сыворотки крови в мг/дл;
- "fbs" = уровень сахара в крови натощак > 120 мг/дл (1 = верно; 0 = ложная повторная);
- "restecg" = ЭКГ в покое (0: норма, 1: наличие аномалии ST-T (инверсии Т-волны и/или повышение или понижение ST > 0,05 мВ), 2: вероятная или определенная гипертрофия левого желудочка по критериям Эстеса);
- "thalach" = максимальная достигнутая частота сердечных сокращений;    
- "exang" = стенокардия, вызванная физической нагрузкой (1 = да, 0 = нет);      
- "oldpeak" = депрессия ST, вызванная физической нагрузкой по сравнению с покоем.
- "slope" = наклон пикового сегмента ST при физической нагрузке (0: восходящий, 1: плоский, 2: нисходящий);      
- "ca" = количество крупных сосудов (0-3), окрашенных при флоуроскопии;        
- "thal" = 0 = ошибка (в исходном наборе данных 0 соответствует NaN), 1 = фиксированный дефект, 2 = норма 3 = обратимый дефект;
- "target" = (0 -нет заболевания, 1 = заболевание).  

Над датасетом работали: 
- Hungarian Institute of Cardiology. Budapest: Andras Janosi, M.D.
- University Hospital, Zurich, Switzerland: William Steinbrunn, M.D.
- University Hospital, Basel, Switzerland: Matthias Pfisterer, M.D.
- V.A. Medical Center, Long Beach and Cleveland Clinic Foundation: Robert Detrano, M.D., PhD.

Ссылка на датасет: https://www.kaggle.com/datasets/data855/heart-disease 

<h3><center>3. Выполнить над датафреймом следующие операции:</center></h3>

- .head().
- .describe().
- Считывание значения конкретной ячейки (с конкретным индексом из конкретной колонки).
- Фильтрация строк по диапазону индекса.
- Фильтрация набора данных по какому-либо условию.
- Работа с пропущенными значениями:
    - удаление строк с пропущенными значениями;
    - заполнение пропущенных значений средним значением по колонке.
    
Если пропущенных значений нет - намеренно их сгенерировать, прибить какие-то куски данных в np.nan.
- Создание нового поля, вычисленного на основе значений других полей:
    - через выражение на базе имеющихся колонок;
    - через DataFrame.apply;
    - через Series.apply.
- Сортировка по какому-либо из полей.
- Вычисление нескольких статистик по колонкам (использовать встроенные агрегатные функции - любые на выбор).
- Вывод по какому-либо полю / набору полей числа значений с использованием .value_counts().
- Вывод уникальных значений какой-либо колонки с использованием .unique().
- Удаление текущего индекса, и создание нового индекса на базе новой колонки, которая для этого лучше всего подходит - см. reset_index и set_index.

Ссылка на датасет: https://www.kaggle.com/datasets/data855/heart-disease

In [6]:
df.head(10) 

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   
5   57    1   0       140   192    0        1      148      0      0.4      1   
6   56    0   1       140   294    0        0      153      0      1.3      1   
7   44    1   1       120   263    0        1      173      0      0.0      2   
8   52    1   2       172   199    1        1      162      0      0.5      2   
9   57    1   2       150   168    0        1      174      0      1.6      2   

    ca  thal  target  
0  0.0   1.0       1  
1  0.0   2.0       1  
2  0.0   2.0       1  
3  0.0   2.0       1  
4  0.0   2.0       1  
5  0.0   1.0       1  
6  0.0   2.0       1  
7  0.0   3.0       1  
8  0.0   3.0       1  
9  0.0   2.0       1

In [7]:
df.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  298.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.674497   
std      0.525860   22.905161    0.469794    1.161075    0.616226    0.938202   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    3.000000   

             thal      target  
count  301.000000  303.000000  
mean     2.328904    0.544554  
std      0.584347    0.498835  
min      1.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

<h4><center>Считывание значения конкретной ячейки (с конкретным индексом из конкретной колонки)</center></h4>

In [8]:
#Считывание значение ячейки под индексом 11 из колонки 'sex'
print(df['sex'][11])

0


<h4><center>Фильтрация строк по диапазону индекса</center></h4>

In [9]:
#Фильтрация строк по диапазону индекса от 10 до 20, при этом добавлена фильттрация столбцов, чтобы визуально было симпотичнее.
print(df.iloc[10:20, 1:4])

    sex  cp  trestbps
10    1   0       140
11    0   2       130
12    1   1       130
13    1   3       110
14    0   3       150
15    0   2       120
16    0   2       120
17    0   3       150
18    1   0       150
19    0   3       140


<h4><center>Фильтрация набора данных по какому-либо условию</center></h4>

In [10]:
#Выводим показатели всех мужчин
df[df['sex'] == 1]

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
3     56    1   1       120   236    0        1      178      0      0.8   
5     57    1   0       140   192    0        1      148      0      0.4   
7     44    1   1       120   263    0        1      173      0      0.0   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
295   63    1   0       140   187    0        0      144      1      4.0   
297   59    1   0       164   176    1        0       90      0      1.0   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   

     slope   ca  thal  target  
0        0  0.0   1.0       1  
1        0  0.0   2.0       1  
3        2  0.0   2.0       1  
5        1  0.0   1.0       1  
7        2  0.0   3.0       1  
..     ...  ...   ...     ...  
295      2  2.0   3.0       0  
297      1  2.0   1.0       0  
299      1  0.0   3.0       0  
300      1  2.0   3.0       0  
301      1  1.0   3.0       0  

[207 rows x 14 columns]

<h4><center>Работа с пропущенными значениями:</center></h4>

- удаление строк с пропущенными значениями

In [11]:
#Тут я просто смотрела под какими индексами Nan
#df[df['ca'].isna()].index
#df[df['thal'].isna()].index

In [12]:
df_no_NaN = df.dropna()

In [13]:
df_no_NaN.info()

<class 'pandas.core.frame.DataFrame'>
Index: 296 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       296 non-null    int64  
 1   sex       296 non-null    int64  
 2   cp        296 non-null    int64  
 3   trestbps  296 non-null    int64  
 4   chol      296 non-null    int64  
 5   fbs       296 non-null    int64  
 6   restecg   296 non-null    int64  
 7   thalach   296 non-null    int64  
 8   exang     296 non-null    int64  
 9   oldpeak   296 non-null    float64
 10  slope     296 non-null    int64  
 11  ca        296 non-null    float64
 12  thal      296 non-null    float64
 13  target    296 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 34.7 KB


*Видно, что удалилось 7 строк.*

- заполнение пропущенных значений средним значением по колонке.

In [14]:
df_no_NaN_mean = df.fillna(df.mean())

In [15]:
df_no_NaN_mean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    float64
 12  thal      303 non-null    float64
 13  target    303 non-null    int64  
dtypes: float64(3), int64(11)
memory usage: 33.3 KB


<h4><center>Создание нового поля, вычисленного на основе значений других полей:</center></h4>

- через выражение на базе имеющихся колонок;
- через DataFrame.apply;
- через Series.apply.

*Я подумала, что нужно сделать что-то логичное. У меня такие данные, что просто друг с другом не проссумируешь значения из колонок (потому что результат будет белеберда), поэтому я создала столбец "Риск гипертонии", который будет заполняться, если у пациента высокий холестин и давление.*

- "trestbps" = кровяное давление в состоянии покоя (в мм рт. ст. при поступлении в больницу);
- "chol" = холестерин сыворотки крови в мг/дл;

In [16]:
print(df['trestbps'].unique())
print(df['chol'].unique())

[145 130 120 140 172 150 110 135 160 105 125 142 155 104 138 128 108 134
 122 115 118 100 124  94 112 102 152 101 132 148 178 129 180 136 126 106
 156 170 146 117 200 165 174 192 144 123 154 114 164]
[233 250 204 236 354 192 294 263 199 168 239 275 266 211 283 219 340 226
 247 234 243 302 212 175 417 197 198 177 273 213 304 232 269 360 308 245
 208 264 321 325 235 257 216 256 231 141 252 201 222 260 182 303 265 309
 186 203 183 220 209 258 227 261 221 205 240 318 298 564 277 214 248 255
 207 223 288 160 394 315 246 244 270 195 196 254 126 313 262 215 193 271
 268 267 210 295 306 178 242 180 228 149 278 253 342 157 286 229 284 224
 206 167 230 335 276 353 225 330 290 172 305 188 282 185 326 274 164 307
 249 341 407 217 174 281 289 322 299 300 293 184 409 259 200 327 237 218
 319 166 311 169 187 176 241 131]


*Возьмем пороговые значения давления = 150 и холестерина = 200 (измерение проводится в  мг/дл, а не как в РФ положено ммоль/л!)*

In [17]:
#Создание нового поля, вычисленного на основе значений других полей через выражение на базе имеющихся колонок
df_copy1 = df.copy() #Создаю копию, чтобы не портить датасет 
df_copy1['Risk of hypertension'] = [(
    'yes' if (chol > 200 and trestbps > 150) else 'no'
) for chol, trestbps in zip(df['chol'], df['trestbps'])]

print(df_copy1)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target Risk of hypertension  
0        0  0.0   1.0       1     

In [18]:
#Создание нового поля, вычисленного на основе значений других полей через DataFrame.apply

df_copy2 = df.copy()

def Roh(row): 
    chol = row['chol']
    trestbps = row['trestbps']
    
    if chol > 200 and trestbps > 150:
        return 'yes'
    else:
        return 'no'

df_copy2['Risk of hypertension'] = df_copy2.apply(Roh, axis=1)

print(df_copy2)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target Risk of hypertension  
0        0  0.0   1.0       1     

In [19]:
#Создание нового поля, вычисленного на основе значений других полей через Series.apply.

df_copy3 = df.copy()

def Roh2(chol, trestbps):
    if chol > 200 and trestbps > 150:
        return 'yes'
    else:
        return 'no'

df_copy3['Risk of hypertension'] = df.apply(lambda row: Roh2(row['chol'], row['trestbps']), axis=1)

print(df_copy3)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target Risk of hypertension  
0        0  0.0   1.0       1     

<h4><center>Сортировка по какому-либо из полей</center></h4>

In [20]:
#Расположим всех пациентов в порядке возрастания давления (колонка 'trestbps')

df_sorted = df.sort_values(by='trestbps')

print(df_sorted)

     age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
71    51    1   2        94   227    0        1      154      1      0.0   
124   39    0   2        94   199    0        1      179      0      0.0   
273   58    1   0       100   234    0        1      156      0      0.1   
66    51    1   2       100   222    0        1      143      1      1.2   
253   67    1   0       100   299    0        0      125      1      0.9   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
203   68    1   2       180   274    1        0      150      1      1.6   
266   55    0   0       180   327    0        2      117      1      3.4   
110   64    0   0       180   325    0        1      154      1      0.0   
248   54    1   1       192   283    0        0      195      0      0.0   
223   56    0   0       200   288    1        0      133      1      4.0   

     slope   ca  thal  target  
71       2  1.0   3.0       1  
124      2  0.0   2.0  

<h4><center>Вычисление нескольких статистик по колонкам (использовать встроенные агрегатные функции - любые на выбор)</center></h4>

In [21]:
#Минимальное и максимальное значение уровня холестерина 
print(df['chol'].agg(['min', 'max']))

min    126
max    564
Name: chol, dtype: int64


<h4><center>Вывод по какому-либо полю/набору полей числа значений с использованием .value_counts()</center></h4>

In [22]:
print(df['thal'].value_counts())

thal
2.0    166
3.0    117
1.0     18
Name: count, dtype: int64


<h4><center>Вывод уникальных значений какой-либо колонки с использованием .unique()</center></h4>

In [23]:
print(df['trestbps'].unique())

[145 130 120 140 172 150 110 135 160 105 125 142 155 104 138 128 108 134
 122 115 118 100 124  94 112 102 152 101 132 148 178 129 180 136 126 106
 156 170 146 117 200 165 174 192 144 123 154 114 164]


<h4><center>Удаление текущего индекса, и создание нового индекса на базе новой колонки, которая для этого лучше всего подходит - см. reset_index и set_index.</center></h4>

In [24]:
#Создается столбец с индексами, которые были у пациентов до set_index
df_reset = df.reset_index(drop=False, inplace=False)
print(df_reset) 

     index  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  \
0        0   63    1   3       145   233    1        0      150      0   
1        1   37    1   2       130   250    0        1      187      0   
2        2   41    0   1       130   204    0        0      172      0   
3        3   56    1   1       120   236    0        1      178      0   
4        4   57    0   0       120   354    0        1      163      1   
..     ...  ...  ...  ..       ...   ...  ...      ...      ...    ...   
298    298   57    0   0       140   241    0        1      123      1   
299    299   45    1   3       110   264    0        1      132      0   
300    300   68    1   0       144   193    1        1      141      0   
301    301   57    1   0       130   131    0        1      115      1   
302    302   57    0   1       130   236    0        0      174      0   

     oldpeak  slope   ca  thal  target  
0        2.3      0  0.0   1.0       1  
1        3.5      0  0.0   2.

In [25]:
#Создание нового индекса на базе колонки 'chol' (не знаю даже, какая колонка может подойти в качестве нового индекса..)
df_new_index = df_reset.set_index('chol')
print(df_new_index)

      index  age  sex  cp  trestbps  fbs  restecg  thalach  exang  oldpeak  \
chol                                                                         
233       0   63    1   3       145    1        0      150      0      2.3   
250       1   37    1   2       130    0        1      187      0      3.5   
204       2   41    0   1       130    0        0      172      0      1.4   
236       3   56    1   1       120    0        1      178      0      0.8   
354       4   57    0   0       120    0        1      163      1      0.6   
...     ...  ...  ...  ..       ...  ...      ...      ...    ...      ...   
241     298   57    0   0       140    0        1      123      1      0.2   
264     299   45    1   3       110    0        1      132      0      1.2   
193     300   68    1   0       144    1        1      141      0      3.4   
131     301   57    1   0       130    0        1      115      1      1.2   
236     302   57    0   1       130    0        0      174      

<h3><center>4. Продемонстрировать работу .groupby, на основе группировок в groupby вычислить агрегатные функции по одной или нескольким колонкам</center></h3>

In [26]:
#Сгруппируем данные по уровню холестерина
dfgby = df.groupby('chol')

In [27]:
type(dfgby)

pandas.core.groupby.generic.DataFrameGroupBy

In [28]:
len(dfgby) #Уникальных значений уровня холестерина 152

152

In [29]:
print(dfgby['trestbps'].mean()) #Среднее значение давления для каждого уровня холестерина
print(dfgby['trestbps'].min()) #Минимальное значение давления для каждого уровня холестерина

chol
126    150.0
131    130.0
141    108.0
149    115.0
157    120.0
       ...  
394    140.0
407    150.0
409    134.0
417    140.0
564    115.0
Name: trestbps, Length: 152, dtype: float64
chol
126    150
131    130
141    108
149    112
157    120
      ... 
394    140
407    150
409    134
417    140
564    115
Name: trestbps, Length: 152, dtype: int64




<h3><center>5. Решейпинг данных 1D -> 2D с использованием .pivot (можно подать на вход результаты агрегатов, полученных ранее с использованием .groupby (сгруппировать по двум полям).</center></h3>

In [30]:
#Группируем по столбцам 'chol', 'cp' и смотрим минимальное и среднее значение давления 
df_grouped = df.groupby(['chol', 'cp'])['trestbps'].agg(['mean', 'min'])

In [31]:
print(df_grouped)

          mean  min
chol cp            
126  2   150.0  150
131  0   130.0  130
141  2   108.0  108
149  0   112.0  112
     2   118.0  118
...        ...  ...
394  0   140.0  140
407  0   150.0  150
409  0   134.0  134
417  2   140.0  140
564  2   115.0  115

[241 rows x 2 columns]


In [32]:
df_pivot = df_grouped.reset_index().pivot(index='chol', columns='cp')
print(df_pivot)

       mean                      min                  
cp        0      1      2   3      0      1      2   3
chol                                                  
126     NaN    NaN  150.0 NaN    NaN    NaN  150.0 NaN
131   130.0    NaN    NaN NaN  130.0    NaN    NaN NaN
141     NaN    NaN  108.0 NaN    NaN    NaN  108.0 NaN
149   112.0    NaN  118.0 NaN  112.0    NaN  118.0 NaN
157     NaN  120.0    NaN NaN    NaN  120.0    NaN NaN
...     ...    ...    ...  ..    ...    ...    ...  ..
394   140.0    NaN    NaN NaN  140.0    NaN    NaN NaN
407   150.0    NaN    NaN NaN  150.0    NaN    NaN NaN
409   134.0    NaN    NaN NaN  134.0    NaN    NaN NaN
417     NaN    NaN  140.0 NaN    NaN    NaN  140.0 NaN
564     NaN    NaN  115.0 NaN    NaN    NaN  115.0 NaN

[152 rows x 8 columns]


<h3><center>6. Решейпинг 1D -> 2D данных, соединённых с группировкой / агрегацией (одним словом - сводная таблица): .pivot_table. Группировать только по категориальным полям или числовым, если значений немного.<h3><center>

In [33]:
df_pivot_table = df.pivot_table(values=['age', 'trestbps', 'chol', 'thalach', 'oldpeak', 'ca'],
                             index=['sex', 'cp'],
                             aggfunc='mean')

In [34]:
df_pivot_table

age        ca        chol   oldpeak     thalach    trestbps
sex cp                                                                   
0   0   57.256410  0.820513  267.538462  1.356410  145.282051  138.589744
    1   51.944444  0.555556  251.444444  0.461111  162.833333  128.055556
    2   54.971429  0.257143  261.057143  0.474286  151.800000  127.885714
    3   63.250000  0.500000  247.000000  1.575000  149.500000  147.500000
1   0   55.105769  0.990291  243.605769  1.393269  138.759615  129.557692
    1   51.031250  0.225806  241.031250  0.234375  162.187500  128.593750
    2   52.538462  0.612245  231.134615  1.015385  158.173077  132.057692
    3   54.315789  0.473684  235.052632  1.352632  157.315789  139.473684

<h3><center>7. Вычислить квантили распределения какого-либо вещественного признака (с использованием numpy.quantile или numpy.percentile).<h3><center>

In [35]:
qua = np.quantile(df['chol'], [0.25, 0.5, 0.75])
print(f'Первый квартиль: {qua[0]}, Второй квартиль: {qua[1]}, Третий квартиль: {qua[2]}')

Первый квартиль: 211.0, Второй квартиль: 240.0, Третий квартиль: 274.5


<h3><center>8. Вычислить (в виде текста) гистограмму какого-либо вещественного признака (с использованием numpy.histogram). Значения гистограммы можно использовать в качестве загрубленного числового признака для заданий (5) или (6).<h3><center>

In [36]:
np.histogram(df["chol"], bins = 7)

(array([ 32, 145, 100,  21,   4,   0,   1], dtype=int64),
 array([126.        , 188.57142857, 251.14285714, 313.71428571,
        376.28571429, 438.85714286, 501.42857143, 564.        ]))

<h3><center>9. Проитерировать DataFrame построчно .iterrows() и выполнить какую-либо операцию внутри цикла.<h3><center>

In [37]:
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target  
0        0  0.0   1.0       1  
1        0  0.0   2.0       1  
2        2  0.0   2.0       1  
3        2  0.0   2.0       1  
4        2  0.0   2.0       1  
..     ...  ...   ...     ...  
298      1  0.0   3.0       0  
299      1  0.0   3.0       0  
300      1  2.0   3.0       0  
301      1  1.0   3.0       0  
302      1  1.0   2.0       0  

[303 rows x 14 columns]

Будем создавать новый столбец, в котором будут указываться категории уровня холестерина пациентов.

In [38]:
df['chol_category'] = ''

In [39]:
for index, row in df.iterrows():
    if row['chol'] < 200:
        df.at[index, 'chol_category'] = 'Low'
    elif 200 <= row['chol'] < 240:
        df.at[index, 'chol_category'] = 'Medium'
    else:
        df.at[index, 'chol_category'] = 'High'

In [40]:
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target chol_category  
0        0  0.0   1.0       1        Medium  
1        0  0.0   2.0       1          High  
2        2  0.0   2.0       1        Medium  
3        2  0.0   2.0       1        Medium  
4        2  0.0   2.0       1          High  
..     ...  ...   ...     ...           ...  
298      1  0.0   3.0       0          High  
299      1  0.0   3.0       0          High  
300      1  2.0   3.0       0           Low  
301      1  1.0   3.0       0           Low  
302      1  1.0   2.0       0        Medium  

[303 rows x 15 columns]

<h3><center>10. Получить DataFrame с MultiIndex любым способом: через конструктор (в документации есть множество видов конструкторов для создания MultiIndex с нуля), через read_sql / read_csv / read_excel, read_*, через pivot_table, через groupby или иными способами.<h3><center>

df_pivot_table похож на DataFrame с MultiIndex

In [41]:
df_multi = df_pivot_table.copy()
df_multi

age        ca        chol   oldpeak     thalach    trestbps
sex cp                                                                   
0   0   57.256410  0.820513  267.538462  1.356410  145.282051  138.589744
    1   51.944444  0.555556  251.444444  0.461111  162.833333  128.055556
    2   54.971429  0.257143  261.057143  0.474286  151.800000  127.885714
    3   63.250000  0.500000  247.000000  1.575000  149.500000  147.500000
1   0   55.105769  0.990291  243.605769  1.393269  138.759615  129.557692
    1   51.031250  0.225806  241.031250  0.234375  162.187500  128.593750
    2   52.538462  0.612245  231.134615  1.015385  158.173077  132.057692
    3   54.315789  0.473684  235.052632  1.352632  157.315789  139.473684

<h4><center>Переставить местами уровни индекса.<h4><center>

In [42]:
df_multi.swaplevel(i=0, j=1, axis = 0).sort_index()

age        ca        chol   oldpeak     thalach    trestbps
cp sex                                                                   
0  0    57.256410  0.820513  267.538462  1.356410  145.282051  138.589744
   1    55.105769  0.990291  243.605769  1.393269  138.759615  129.557692
1  0    51.944444  0.555556  251.444444  0.461111  162.833333  128.055556
   1    51.031250  0.225806  241.031250  0.234375  162.187500  128.593750
2  0    54.971429  0.257143  261.057143  0.474286  151.800000  127.885714
   1    52.538462  0.612245  231.134615  1.015385  158.173077  132.057692
3  0    63.250000  0.500000  247.000000  1.575000  149.500000  147.500000
   1    54.315789  0.473684  235.052632  1.352632  157.315789  139.473684

<h4><center>Транспонировать таблицу (или создать новую другую) с MultiIndex.<h4><center>

In [43]:
df_multi.T

sex                0                                            1              \
cp                 0           1           2        3           0           1   
age        57.256410   51.944444   54.971429   63.250   55.105769   51.031250   
ca          0.820513    0.555556    0.257143    0.500    0.990291    0.225806   
chol      267.538462  251.444444  261.057143  247.000  243.605769  241.031250   
oldpeak     1.356410    0.461111    0.474286    1.575    1.393269    0.234375   
thalach   145.282051  162.833333  151.800000  149.500  138.759615  162.187500   
trestbps  138.589744  128.055556  127.885714  147.500  129.557692  128.593750   

sex                               
cp                 2           3  
age        52.538462   54.315789  
ca          0.612245    0.473684  
chol      231.134615  235.052632  
oldpeak     1.015385    1.352632  
thalach   158.173077  157.315789  
trestbps  132.057692  139.473684

<h4><center>Удалить один из уровней индекса или добавить новый уровень индекса (можно инициализированный константой) - см. документацию.<h4><center>

In [44]:
df_reset = df_multi.reset_index(level='cp')
df_reset

cp        age        ca        chol   oldpeak     thalach    trestbps
sex                                                                       
0     0  57.256410  0.820513  267.538462  1.356410  145.282051  138.589744
0     1  51.944444  0.555556  251.444444  0.461111  162.833333  128.055556
0     2  54.971429  0.257143  261.057143  0.474286  151.800000  127.885714
0     3  63.250000  0.500000  247.000000  1.575000  149.500000  147.500000
1     0  55.105769  0.990291  243.605769  1.393269  138.759615  129.557692
1     1  51.031250  0.225806  241.031250  0.234375  162.187500  128.593750
1     2  52.538462  0.612245  231.134615  1.015385  158.173077  132.057692
1     3  54.315789  0.473684  235.052632  1.352632  157.315789  139.473684

<h3><center>11. Продемонстировать работу .merge.<h3><center>

In [45]:
df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target chol_category  
0        0  0.0   1.0       1        Medium  
1        0  0.0   2.0       1          High  
2        2  0.0   2.0       1        Medium  
3        2  0.0   2.0       1        Medium  
4        2  0.0   2.0       1          High  
..     ...  ...   ...     ...           ...  
298      1  0.0   3.0       0          High  
299      1  0.0   3.0       0          High  
300      1  2.0   3.0       0           Low  
301      1  1.0   3.0       0           Low  
302      1  1.0   2.0       0        Medium  

[303 rows x 15 columns]

In [47]:
df_redact = pd.DataFrame({
    "age" : df["age"],
    "cholesterine" : df["chol"],
    "Tha": df["thalach"]
})

In [49]:
merged_df = pd.merge(df, df_redact, left_index=True, right_index=True)

In [50]:
merged_df

age_x  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0       63    1   3       145   233    1        0      150      0      2.3   
1       37    1   2       130   250    0        1      187      0      3.5   
2       41    0   1       130   204    0        0      172      0      1.4   
3       56    1   1       120   236    0        1      178      0      0.8   
4       57    0   0       120   354    0        1      163      1      0.6   
..     ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298     57    0   0       140   241    0        1      123      1      0.2   
299     45    1   3       110   264    0        1      132      0      1.2   
300     68    1   0       144   193    1        1      141      0      3.4   
301     57    1   0       130   131    0        1      115      1      1.2   
302     57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target chol_category  age_y  cholesterine  Tha  
0        0  0.0   1.0       1        Medium     63           233  150  
1        0  0.0   2.0       1          High     37           250  187  
2        2  0.0   2.0       1        Medium     41           204  172  
3        2  0.0   2.0       1        Medium     56           236  178  
4        2  0.0   2.0       1          High     57           354  163  
..     ...  ...   ...     ...           ...    ...           ...  ...  
298      1  0.0   3.0       0          High     57           241  123  
299      1  0.0   3.0       0          High     45           264  132  
300      1  2.0   3.0       0           Low     68           193  141  
301      1  1.0   3.0       0           Low     57           131  115  
302      1  1.0   2.0       0        Medium     57           236  174  

[303 rows x 18 columns]

<h3><center>12. Продемонстрировать работу с .concat или append.<h3><center>

In [56]:
df_part1 = df.iloc[0:10]  
df_part2 = df.iloc[290:]  

concatenated_df = pd.concat([df_part1, df_part2], axis=0)

In [57]:
concatenated_df

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
5     57    1   0       140   192    0        1      148      0      0.4   
6     56    0   1       140   294    0        0      153      0      1.3   
7     44    1   1       120   263    0        1      173      0      0.0   
8     52    1   2       172   199    1        1      162      0      0.5   
9     57    1   2       150   168    0        1      174      0      1.6   
290   61    1   0       148   203    0        1      161      0      0.0   
291   58    1   0       114   318    0        2      140      0      4.4   
292   58    0   0       170   225    1        0      146      1      2.8   
293   67    1   2       152   212    0        0      150      0      0.8   
294   44    1   0       120   169    0        1      144      1      2.8   
295   63    1   0       140   187    0        0      144      1      4.0   
296   63    0   0       124   197    0        1      136      1      0.0   
297   59    1   0       164   176    1        0       90      0      1.0   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope   ca  thal  target chol_category  
0        0  0.0   1.0       1        Medium  
1        0  0.0   2.0       1          High  
2        2  0.0   2.0       1        Medium  
3        2  0.0   2.0       1        Medium  
4        2  0.0   2.0       1          High  
5        1  0.0   1.0       1           Low  
6        1  0.0   2.0       1          High  
7        2  0.0   3.0       1          High  
8        2  0.0   3.0       1           Low  
9        2  0.0   2.0       1           Low  
290      2  1.0   3.0       0        Medium  
291      0  3.0   1.0       0          High  
292      1  2.0   1.0       0        Medium  
293      1  0.0   3.0       0        Medium  
294      0  0.0   1.0       0           Low  
295      2  2.0   3.0       0           Low  
296      1  0.0   2.0       0           Low  
297      1  2.0   1.0       0           Low  
298      1  0.0   3.0       0          High  
299      1  0.0   3.0       0          High  
300      1  2.0   3.0       0           Low  
301      1  1.0   3.0       0           Low  
302      1  1.0   2.0       0        Medium